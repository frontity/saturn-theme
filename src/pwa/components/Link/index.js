/* global window */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import NProgress from 'nprogress';
import { inject } from 'mobx-react';

class Link extends Component {
  static propTypes = {
    type: PropTypes.string.isRequired,
    id: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
    page: PropTypes.number,
    context: PropTypes.shape({}),
    method: PropTypes.string,
    eventCategory: PropTypes.string,
    eventAction: PropTypes.string,
    children: PropTypes.node.isRequired,
    href: PropTypes.string.isRequired,
    routeChangeRequested: PropTypes.func.isRequired,
    sendEvent: PropTypes.func.isRequired,
  };

  static defaultProps = {
    page: null,
    method: 'push',
    context: null,
    eventCategory: null,
    eventAction: null,
  };

  constructor() {
    super();

    this.linkClicked = this.linkClicked.bind(this);
  }

  linkClicked(e) {
    // ignore click for new tab / new window behavior
    if (
      e.currentTarget.nodeName === 'A' &&
      (e.metaKey ||
        e.ctrlKey ||
        e.shiftKey ||
        (e.nativeEvent && e.nativeEvent.which === 2))
    )
      return;
    e.preventDefault();
    NProgress.configure({ showSpinner: false });
    NProgress.start();

    const {
      routeChangeRequested,
      sendEvent,
      type,
      id,
      page,
      context,
      method,
      eventCategory,
      eventAction,
    } = this.props;
    setTimeout(() => {
      sendEvent({ category: eventCategory, action: eventAction });

      routeChangeRequested({
        selectedItem: { type, id, page },
        context,
        method,
      });
    }, 100);
  }

  render() {
    const { children, href } = this.props;
    return React.cloneElement(children, { onClick: this.linkClicked, href });
  }
}

export default inject(
  ({ stores: { connection, analytics } }, { type, id, page }) => ({
    href: page
      ? connection.entity(type, id).pagedLink(page)
      : connection.entity(type, id).link,
    routeChangeRequested: connection.routeChangeRequested,
    sendEvent: analytics.sendEvent,
  }),
)(Link);
