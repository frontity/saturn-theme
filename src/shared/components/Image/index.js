/* eslint-disable jsx-a11y/anchor-is-valid, react/no-danger */
import React, { Component } from 'react';
import { bool, string } from 'prop-types';
import { inject } from 'mobx-react';
import styled from 'styled-components';
import IconImage from '../Icons/Image';

class Image extends Component {
  static propTypes = {
    isContent: bool,
    width: string, // CSS values
    height: string, // CSS values
    alt: string,
    src: string,
    srcSet: string,
    sizes: string,
    isAmp: bool.isRequired,
    hasPlaceholder: bool,
    objectFit: string,
  };

  static defaultProps = {
    isContent: false,
    width: 'auto',
    height: 'auto',
    alt: null,
    src: null,
    srcSet: null,
    sizes: null,
    hasPlaceholder: true,
    objectFit: 'cover',
  };

  componentDidMount() {
    window.document.lazyLoadInstance.update();
  }

  render() {
    const {
      alt,
      width,
      height,
      isContent,
      src,
      srcSet,
      sizes,
      isAmp,
      hasPlaceholder,
      objectFit,
    } = this.props;

    if (isAmp) {
      return (
        <Container isContent={isContent} styles={{ height, width }}>
          {src || srcSet ? (
            <amp-img
              alt={alt}
              src={src}
              srcSet={srcSet}
              sizes={sizes}
              layout="fill"
            />
          ) : null}
        </Container>
      );
    }

    return (
      <Container
        isContent={isContent}
        objectFit={objectFit}
        styles={{ height, width }}
      >
        {hasPlaceholder && (
          <Icon isContent={isContent} styles={{ height, width }}>
            <IconImage size={40} />
          </Icon>
        )}
        <span
          suppressHydrationWarning
          dangerouslySetInnerHTML={{
            __html:
              src || srcSet
                ? `<img
                    alt="alt"
                    class="lazy"
                    ${src && `data-src="${src}"`}
                    ${srcSet && `data-srcset="${srcSet}"`}
                    ${sizes && `data-sizes="${sizes}"`}
                  >`
                : null,
          }}
        />
      </Container>
    );
  }
}

export default inject(
  ({ stores: { connection, build } }, { id, isContent, width, height }) => {
    if (!id) return { isAmp: build.isAmp };

    const media = connection.entity('media', id);

    return {
      id,
      isAmp: build.isAmp,
      isContent: !!isContent,
      alt: media.alt,
      src: media.src,
      srcSet: media.srcSet,
      sizes: media.srcSet && `${parseInt(width, 10) || 100}vw`,
      width: width || '100vw',
      height:
        height ||
        `${100 * (media.original.height / media.original.width || 0.6)}vw`,
    };
  },
)(Image);

const Container = styled.span`
  display: ${({ isContent }) => (isContent ? 'block' : 'flex')};
  ${({ isContent }) => (isContent ? '' : 'align-items: stretch')};
  box-sizing: border-box;
  width: ${({ styles }) => styles.width};
  height: ${({ styles }) => styles.height};
  position: relative;
  margin: ${({ isContent }) => (isContent ? '15px 0' : '')};
  ${({ isContent }) => isContent && 'left: -15px'};

  span > img {
    ${({ isContent, styles }) =>
      isContent && styles.height === 'auto'
        ? 'position: static'
        : 'position: absolute'};
    width: 100%;
    height: 100%;
    object-fit: ${({ objectFit }) => objectFit};
    object-position: center;
    color: transparent;
    border: none;
    opacity: 0;
    transition: opacity ${({ theme }) => theme.transitionTime} ease-in;
  }

  amp-img > img {
    object-fit: cover;
  }

  img.loaded {
    opacity: 1;
  }
`;

const Icon = styled.span`
  position: absolute;
  top: 0;
  color: #bdbdbd;
  width: 100%;
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  ${({ isContent, styles }) =>
    isContent && styles.height === 'auto' ? 'z-index: -1' : ''};
`;
