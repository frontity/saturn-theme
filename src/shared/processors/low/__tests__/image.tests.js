/* eslint-disable react/prop-types */
import Image from '../../../components/Image';
import Link from '../../../../pwa/components/Link';
import processor from '../image';

const imageWithId = {
  type: 'element',
  component: 'img',
  props: {
    src: 'media/image.jpg',
    srcset: 'media/image.jpg',
    alt: 'image 1',
    'data-attachment-id': '23',
  },
};

const imageWithoutSrc = {
  type: 'element',
  component: 'img',
  props: {
    alt: 'image 1',
  },
};

const imageWithoutId = {
  type: 'element',
  component: 'img',
  props: {
    src: '/media/image.jpg',
    srcSet: '/media/image.jpg',
    alt: 'image 1',
  },
};

const imageWithDataSrc = {
  type: 'element',
  component: 'img',
  props: {
    src:
      'https://frontity.com/wp-content/plugins/media-ace/includes/lazy-load/images/blank.png',
    'data-src': 'https://frontity.com/wp-content/uploads/2018/10/image.jpg',
    'data-srcset':
      'https://frontity.com/wp-content/uploads/2018/10/image.jpg 800w, https://frontity.com/wp-content/uploads/2018/10/image-300x225.jpg 300w, https://frontity.com/wp-content/uploads/2018/10/image-768x576.jpg 768w, https://frontity.com/wp-content/uploads/2018/10/image-192x144.jpg 192w, https://frontity.com/wp-content/uploads/2018/10/image-384x288.jpg 384w, https://frontity.com/wp-content/uploads/2018/10/image-90x68.jpg 90w, https://frontity.com/wp-content/uploads/2018/10/image-180x135.jpg 180w, https://frontity.com/wp-content/uploads/2018/10/image-561x421.jpg 561w, https://frontity.com/wp-content/uploads/2018/10/image-364x273.jpg 364w, https://frontity.com/wp-content/uploads/2018/10/image-758x569.jpg 758w, https://frontity.com/wp-content/uploads/2018/10/image-608x456.jpg 608w, https://frontity.com/wp-content/uploads/2018/10/image-64x48.jpg 64w, https://frontity.com/wp-content/uploads/2018/10/image-128x96.jpg 128w',
    'data-sizes': '(max-width: 800px) 100vw, 800px',
  },
};

const imageWithDataSrcAndNoHostname = {
  type: 'element',
  component: 'img',
  props: {
    src: '/wp-content/plugins/media-ace/includes/lazy-load/images/blank.png',
    'data-src': '/wp-content/uploads/2018/10/image.jpg',
    'data-srcset':
      '/wp-content/uploads/2018/10/image.jpg 800w, /wp-content/uploads/2018/10/image-300x225.jpg 300w',
  },
};

const imageWithDataOriginal = {
  type: 'element',
  component: 'img',
  props: {
    alt: 'image 1',
    'data-original': '/media/image.jpg',
  },
};

const noImage = {
  type: 'element',
  component: 'div',
  props: { className: 'whatever' },
};

const item = {
  type: 'post',
  id: 60,
  entity: {
    media: {
      featured: {
        id: 23,
      },
      content: [23, 24, 25],
    },
  },
};

const stores = {
  settings: {
    generalSite: {
      url: 'https://test.frontity.io',
    },
  },
};

const payload = { stores, item };

describe('Theme › Processors › image', () => {
  test('does not pass test with invalid elements', () => {
    expect(processor.test(noImage)).toBeFalsy();
  });
  test('processes images that returns src=null and srcSet=null', () => {
    const element = processor.process(imageWithoutSrc, payload);
    expect(element.component).toBe(Image);
    expect(element.props.src).toBeNull();
    expect(element.props.srcSet).toBeNull();
  });
  test('passes test with valid elements', () => {
    expect(processor.test(imageWithId)).toBeTruthy();
    expect(processor.test(imageWithoutId)).toBeTruthy();
    expect(processor.test(imageWithDataOriginal)).toBeTruthy();
  });
  test('processes images with data-attachment-id', () => {
    const element = processor.process(imageWithId, payload);
    expect(element.component).toBe(Link);
    expect(element.children[0].children[0].component).toBe(Image);
    expect(element).toMatchSnapshot();
  });
  test('processes images without data-attachment-id', () => {
    const element = processor.process(imageWithoutId, payload);
    expect(element.component).toBe(Image);
    expect(element).toMatchSnapshot();
  });
  test('processes images with data-original', () => {
    const element = processor.process(imageWithDataOriginal, payload);
    expect(element.component).toBe(Image);
    expect(element).toMatchSnapshot();
  });
  test('process images with data-src', () => {
    const element = processor.process(imageWithDataSrc, payload);
    expect(element.component).toBe(Image);
    expect(element.props.src).toBe(
      'https://frontity.com/wp-content/uploads/2018/10/image.jpg',
    );
    expect(element.props.srcSet).toBe(
      'https://frontity.com/wp-content/uploads/2018/10/image.jpg 800w, https://frontity.com/wp-content/uploads/2018/10/image-300x225.jpg 300w, https://frontity.com/wp-content/uploads/2018/10/image-768x576.jpg 768w, https://frontity.com/wp-content/uploads/2018/10/image-192x144.jpg 192w, https://frontity.com/wp-content/uploads/2018/10/image-384x288.jpg 384w, https://frontity.com/wp-content/uploads/2018/10/image-90x68.jpg 90w, https://frontity.com/wp-content/uploads/2018/10/image-180x135.jpg 180w, https://frontity.com/wp-content/uploads/2018/10/image-561x421.jpg 561w, https://frontity.com/wp-content/uploads/2018/10/image-364x273.jpg 364w, https://frontity.com/wp-content/uploads/2018/10/image-758x569.jpg 758w, https://frontity.com/wp-content/uploads/2018/10/image-608x456.jpg 608w, https://frontity.com/wp-content/uploads/2018/10/image-64x48.jpg 64w, https://frontity.com/wp-content/uploads/2018/10/image-128x96.jpg 128w',
    );
    expect(element.props['data-src']).toBeUndefined();
    expect(element.props['data-srcset']).toBeUndefined();
    expect(element).toMatchSnapshot();
  });
  test('process images with data-src and without hostname', () => {
    const element = processor.process(imageWithDataSrcAndNoHostname, payload);
    expect(element.component).toBe(Image);
    expect(element.props.src).toBe(
      'https://test.frontity.io/wp-content/uploads/2018/10/image.jpg',
    );
    expect(element.props.srcSet).toBe(
      'https://test.frontity.io/wp-content/uploads/2018/10/image.jpg 800w, https://test.frontity.io/wp-content/uploads/2018/10/image-300x225.jpg 300w',
    );
    expect(element.props['data-src']).toBeUndefined();
    expect(element.props['data-srcset']).toBeUndefined();
    expect(element).toMatchSnapshot();
  });
});
