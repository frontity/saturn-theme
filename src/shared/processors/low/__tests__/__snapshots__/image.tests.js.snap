// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Theme › Processors › image process images with data-src 1`] = `
Object {
  "children": null,
  "component": [Function],
  "props": Object {
    "alt": null,
    "height": "auto",
    "isContent": true,
    "key": "https://frontity.com/wp-content/uploads/2018/10/image.jpg",
    "src": "https://frontity.com/wp-content/uploads/2018/10/image.jpg",
    "srcSet": "https://frontity.com/wp-content/uploads/2018/10/image.jpg 800w, https://frontity.com/wp-content/uploads/2018/10/image-300x225.jpg 300w, https://frontity.com/wp-content/uploads/2018/10/image-768x576.jpg 768w, https://frontity.com/wp-content/uploads/2018/10/image-192x144.jpg 192w, https://frontity.com/wp-content/uploads/2018/10/image-384x288.jpg 384w, https://frontity.com/wp-content/uploads/2018/10/image-90x68.jpg 90w, https://frontity.com/wp-content/uploads/2018/10/image-180x135.jpg 180w, https://frontity.com/wp-content/uploads/2018/10/image-561x421.jpg 561w, https://frontity.com/wp-content/uploads/2018/10/image-364x273.jpg 364w, https://frontity.com/wp-content/uploads/2018/10/image-758x569.jpg 758w, https://frontity.com/wp-content/uploads/2018/10/image-608x456.jpg 608w, https://frontity.com/wp-content/uploads/2018/10/image-64x48.jpg 64w, https://frontity.com/wp-content/uploads/2018/10/image-128x96.jpg 128w",
    "width": "100vw",
  },
  "type": "element",
}
`;

exports[`Theme › Processors › image processes images with data-attachment-id 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": null,
          "component": [Function],
          "props": Object {
            "id": 23,
            "isContent": true,
            "key": "23",
          },
          "type": "element",
        },
      ],
      "component": "a",
      "props": Object {},
      "type": "element",
    },
  ],
  "component": [Function],
  "props": Object {
    "context": Object {
      "columns": Array [
        Array [
          Object {
            "id": 23,
            "type": "media",
          },
        ],
        Array [
          Object {
            "id": 24,
            "type": "media",
          },
        ],
        Array [
          Object {
            "id": 25,
            "type": "media",
          },
        ],
      ],
      "options": Object {
        "bar": "media",
      },
    },
    "eventAction": "open content media",
    "eventCategory": "Post",
    "id": 23,
    "type": "media",
  },
  "type": "element",
}
`;

exports[`Theme › Processors › image processes images with data-original 1`] = `
Object {
  "children": null,
  "component": [Function],
  "props": Object {
    "alt": "image 1",
    "height": "auto",
    "isContent": true,
    "key": "https://test.frontity.io/media/image.jpg",
    "src": "https://test.frontity.io/media/image.jpg",
    "srcSet": null,
    "width": "100vw",
  },
  "type": "element",
}
`;

exports[`Theme › Processors › image processes images without data-attachment-id 1`] = `
Object {
  "children": null,
  "component": [Function],
  "props": Object {
    "alt": "image 1",
    "height": "auto",
    "isContent": true,
    "key": "https://test.frontity.io/media/image.jpg",
    "src": "https://test.frontity.io/media/image.jpg",
    "srcSet": "https://test.frontity.io/media/image.jpg",
    "width": "100vw",
  },
  "type": "element",
}
`;
